Weak form modal DG methods on simplicial (and quad/hex meshes)

jlchan opened the pull request:
After reading through @ranocha and @andrewwinters5000 new solvers, I think I have a better idea of how to structure mine. Closed https://github.com/trixi-framework/Trixi.jl/pull/557 in favor of this PR, which I think will be better organized and more readable (hopefully). 

Currently passes the "preserve constant state" test for `CompressibleEulerEquations2D` using a triangular DG discretization. Periodic entropy sine wave seems to work now too.

Todo: 
- [x] Incorporate dissipation. Done, but only supports FluxPlusDissipation types.
- [x] new mesh type (draft interface with my ::MeshData)
- [x] imposing boundary conditions
- [x] add tests

I think I may save flux differencing for another PR so this one is more readable.

jlchan commented:
Can I change this default to `nelements(...)`? https://github.com/trixi-framework/Trixi.jl/blob/547170ac694ea414c51b0c31d0bc70c7681f71a1/src/callbacks_step/analysis.jl#L83. Any general code which assumes structure in `cache` is hard to extend.

sloede commented:
> Can I change this default to `nelements(...)`?
> 
> https://github.com/trixi-framework/Trixi.jl/blob/547170ac694ea414c51b0c31d0bc70c7681f71a1/src/callbacks_step/analysis.jl#L83
> . Any general code which assumes structure in `cache` is hard to extend.

I don't think `nelements(...)` will give you what you want, since this returns the number of elements, while `uEltype` expects the core *data type* in which the solution state is stored. Typically it's `Float64`, but it is different when using, e.g., `Measurements.jl` or `ForwardDiff.jl`. Right now, the only portable way to do this that I can see is to *not* destructure the semi in https://github.com/trixi-framework/Trixi.jl/blob/547170ac694ea414c51b0c31d0bc70c7681f71a1/src/callbacks_step/analysis.jl#L68-L71 but to use something like `eltype(semi)` here. But @ranocha might have a better idea.

jlchan commented:
Ah, yes - I don't want `nelements` (not sure what I was thinking).

I ended up passing in `uEltype = real(dg)` as a keyword argument, though I'm not sure that's the same thing.

sloede commented:
> I ended up passing in `uEltype = real(dg)` as a keyword argument, though I'm not sure that's the same thing.

It is not. E.g., `real(solver)` will virtually always default to `Float64`, while it is e.g., `Measurements{Float64}` when using `Measurements.jl`. Thus if you write it as `real(dg)`, these parts of the code will not support automatic differentiation etc.

jlchan commented:
Thanks. So then if `uEltype` is the solution type, what is `real(solver)` dused for?

sloede commented:
> Thanks. So then if `uEltype` is the solution type, what is `real(solver)` used for?

Whenever a non-mutable real type is required, e.g., for coordinates or physical constants etc. Although this distinction becomes blurred if you want to do shape optimization or similar shenanigans.

jlchan commented:
OK, that makes more sense. I'm used to inferring the types directly from the solution or coordinates, so the `uEltype` and `real(dg)` are new to me. 

The PR is almost ready to review, but the tests I've added don't seem to be using the correct version of the dependent packages. I've added `StartUpDG#master` to `test/Project.toml`, but the CI test is still using `StartUpDG` 0.7.0. Do I need to specify this information elsewhere (like in the CI files)?

jlchan commented:
Alternatively, is there a way to run the tests locally? Just doing `] test Trixi` gives the error `ERROR: can not merge projects`

sloede commented:
> The PR is almost ready to review, but the tests I've added don't seem to be using the correct version of the dependent packages. I've added `StartUpDG#master` to `test/Project.toml`, but the CI test is still using `StartUpDG` 0.7.0. Do I need to specify this information elsewhere (like in the CI files)?

The `Project.toml` file does not specify the source of any package. Thus even though you add it as `#master`, it will just use whatever is in the registry, as far as I know. What we had to do with HOHQMesh.jl to be able to use, e.g., the unregistered HOHQMesh_jll.jl package, was to also commit the `Manifest.toml`.

I'd try the following: Go to your `test/` directory. Remove the current entry of `StartUpDG` from the `Project.toml` and *delete* the `Manifest.toml`. Then run
```bash
julia --project=. -e 'import Pkg; Pkg.add("url/to/StartUpDG.jl#master")'
```
(I might be wrong about the exact `Pkg` syntax, but you get the drift). Finally, commit the `Manifest.toml` and try again with testing.

In addition to the `Manifest.toml` in `test/`, you also might have have to change & commit the `Manifest.toml` in the Trixi root directory accordingly, but I am not so firm on that. It was a lot of trial and error until I got the CI stuff to run properly.

ranocha commented:
> > Can I change this default to `nelements(...)`?
> > https://github.com/trixi-framework/Trixi.jl/blob/547170ac694ea414c51b0c31d0bc70c7681f71a1/src/callbacks_step/analysis.jl#L83
> > 
> > . Any general code which assumes structure in `cache` is hard to extend.
> 
> I don't think `nelements(...)` will give you what you want, since this returns the number of elements, while `uEltype` expects the core _data type_ in which the solution state is stored. Typically it's `Float64`, but it is different when using, e.g., `Measurements.jl` or `ForwardDiff.jl`. Right now, the only portable way to do this that I can see is to _not_ destructure the semi in
> 
> https://github.com/trixi-framework/Trixi.jl/blob/547170ac694ea414c51b0c31d0bc70c7681f71a1/src/callbacks_step/analysis.jl#L68-L71
> 
> but to use something like `eltype(semi)` here. But @ranocha might have a better idea.

Maybe we should have methods like
```julia
real(mesh, equations, solver, cache)
eltype(mesh, equations, solver, cache)
```
that can be overwritten via multiple dispatch? For our current set of meshes, we could have
```julia
function Base.real(mesh::OurCurrentMeshTypes, equations, solver, cache)
  promote_type(real(mesh), real(equations), real(solver))
end
```
and get something from `cache` for `eltype`. Then, you could specialize these methods for your new types, @jlchan.

ranocha commented:
Would it be an option to register a new version of StartUpDg for you, @jlchan?

jlchan commented:
@sloede Thanks - I always forget about the Manifest issues. I'll try that.

@ranocha yes, I can register a new version of StartUpDG. The issue is that I'm also using the master branch of StructArrays. I can ask the devs if they're ok with releasing a new version, but I think they tend to be pretty conservative on this.

jlchan commented:
@ranocha the dispatch-based version of `real` and `uEltype` sound great!

ranocha commented:
> @ranocha yes, I can register a new version of StartUpDG. The issue is that I'm also using the master branch of StructArrays. I can ask the devs if they're ok with releasing a new version, but I think they tend to be pretty conservative on this.

That's an annoying issue...

jlchan commented:
I've registered a new version of StartUpDG and asked the StructArrays dev when there might be a new version. Fingers crossed!

jlchan commented:
Looks like StructArrays will release a new version. Hopefully this fixed our CI issues.

ranocha commented:
> Looks like StructArrays will release a new version [JuliaArrays/StructArrays.jl#192](https://github.com/JuliaArrays/StructArrays.jl/pull/192) when the upstream issue [JuliaData/WeakRefStrings.jl#72](https://github.com/JuliaData/WeakRefStrings.jl/issues/72) gets fixed.
> 
> Should fix our CI issues.

Sounds good :+1: Please request a review (from Michael and me) when you think the PR is basically finished, even if we need to wait a bit more before CI can run due to upstream issues.

jlchan commented:
@ranocha thanks! In that case, the PR is ready for review. 

One side note - I'm a bit unsatisfied with my naming "dg_tri". This isn't really "triangular" DG, since it also supports Quad meshes too. I was thinking "modal" DG, but it also supports triangular SBP operators.

sloede commented:
> One side note - I'm a bit unsatisfied with my naming "dg_tri". This isn't really "triangular" DG, since it also supports Quad meshes too. I was thinking "modal" DG, but it also supports triangular SBP operators.

`dg_onetorulethemall`?

But seriously, if it supports nearly everything, is there maybe another characteristic by which it can be distinguished? Eg, a name like `dg_hesthaven_style` but (much) better than what I can come up with? Or maybe focus on what you plan to use it for - even if supports other modes in theory, maybe you won't use them for now and thus they are not readily supported (yet).

jlchan commented:
That's an interesting observation - it is sort of Kopriva vs Hesthaven/Warburton style, isn't it? 

I'll stick with dg_tri for now, since the Quad stuff works (you can trigger it by replacing `Tri()` with `Quad()` in any of the elixirs) but is only supported for affine quad meshes at the moment. Once I get curved quads implemented, may be worth revisiting naming.

ranocha commented:
Without having read the code of this new PR, I guess that you will still use a kind of global linear algebra style (in contrast to the local loop style that the currently implemented DGSEM solvers use). In addition, it will be able to represent different mesh types. `dg_poly`, `dg_global`, `dg_matmul` or something along these lines?

jlchan commented:
> Without having read the code of this new PR, I guess that you will still use a kind of global linear algebra style (in contrast to the local loop style that the currently implemented DGSEM solvers use). In addition, it will be able to represent different mesh types. `dg_poly`, `dg_global`, `dg_matmul` or something along these lines?

Sort of - most routines are not global in the "entire mesh" sense, just locally across all nodes on each element. Since triangular operators are dense, there's not as much reason to loop over individual nodes IMO. Quads are supported by just building multi-dimensional operators via tensor product, which will be slower than hand-crafted loops.

I think the other significant difference in the current approach is the use of nodal maps - rather than store element orientations, we just store face-to-face connectivity and build index maps between nodes on shared faces. I find orientations confusing to keep track of (esp for more than one element type), and this approach avoids using them.

jlchan commented:
One downside of the index mapping approach to connectivities: we do some extra work since surface fluxes are computed twice (once for each element rather than once per face).

sloede commented:
I think emphasizing the triangular approach is fine for now, at least as long as it is still marked experimental. However, is it strictly 2D what you're doing? Or would it make sense to name it `dg_simplex`?  That would, if course, only make sense if 3D is really just a few lines of code away.

As y'all know, I am not a big fan of abbreviations, thus I'd slightly favor something like `dg_triangles`.

jlchan commented:
> I think emphasizing the triangular approach is fine for now, at least as long as it is still marked experimental. However, is it strictly 2D what you're doing? Or would it make sense to name it `dg_simplex`? That would, if course, only make sense if 3D is really just a few lines of code away.
>
> As y'all know, I am not a big fan of abbreviations, thus I'd slightly favor something like `dg_triangles`.

It will be, though I'm working on tet support in StartUpDG still. For that reason, I'll go ahead and change the folder to dg_triangles for now.

ranocha commented on .github/workflows/ci.yml:
```diff
@@ -58,6 +58,7 @@ jobs:
           - 2d_part1
           - 2d_part2
           - 2d_part3
+          - 2d_tri
```
```suggestion
          - 2d_triangles
```

ranocha commented on examples/triangular_mesh_2D/elixir_ape_triangular_mesh.jl:
```diff
@@ -0,0 +1,45 @@
+using Trixi, OrdinaryDiffEq
+
+rd = RefElemData(Tri(), N=4)
+dg = DG(rd, (), SurfaceIntegralWeakForm(FluxLaxFriedrichs()), VolumeIntegralWeakForm())
```
```suggestion
dg = DG(rd, nothing #= mortar =#, 
        SurfaceIntegralWeakForm(FluxLaxFriedrichs()), VolumeIntegralWeakForm())
```

ranocha commented on examples/triangular_mesh_2D/elixir_ape_triangular_mesh.jl:
```diff
@@ -0,0 +1,45 @@
+                                    source_terms = source_terms, 
+                                    boundary_conditions = boundary_conditions) 
+
+tspan = (0.0, .1)
```
```suggestion
tspan = (0.0, 0.1)
```

ranocha commented on examples/triangular_mesh_2D/elixir_euler_periodic_triangular_mesh.jl:
```diff
@@ -0,0 +1,33 @@
+using Trixi, OrdinaryDiffEq
+
+rd = RefElemData(Tri(), N=4) # equivalent to a "basis"
+dg = DG(rd, (), SurfaceIntegralWeakForm(FluxHLL()), VolumeIntegralWeakForm())
```
```suggestion
dg = DG(rd, nothing #= mortar =#, 
        SurfaceIntegralWeakForm(FluxHLL()), VolumeIntegralWeakForm())
```

ranocha commented on examples/triangular_mesh_2D/elixir_euler_triangular_mesh.jl:
```diff
@@ -0,0 +1,45 @@
+using Trixi, OrdinaryDiffEq
+
+rd = RefElemData(Tri(), N=4)
+dg = DG(rd, (), SurfaceIntegralWeakForm(FluxHLL()), VolumeIntegralWeakForm())
```
```suggestion
dg = DG(rd, nothing #= mortar =#, 
        SurfaceIntegralWeakForm(FluxHLL()), VolumeIntegralWeakForm())
```

ranocha commented on src/solvers/dg_triangles/analysis.jl:
```diff
@@ -0,0 +1,76 @@
+  # to polynomials of degree N and testing with the L2 projection of v(u) 
+  # would be equivalent to testing with v(u) due to the moment-preserving 
+  # property of the L2 projection.
+  dS_dt = zero(real(dg))
```
Should probably be `eltype(du_values)`, since these are usually more general when using AD or something like that.

jlchan replied:
Changed it to `zero(eltype(du[1]))` since `du` is a StructArray of arrays.

ranocha commented on src/solvers/dg_triangles/analysis.jl:
```diff
@@ -0,0 +1,76 @@
+## todo: what else is required in the interface for AnalysisCallback?
+SolutionAnalyzer(rd::RefElemData) = rd
+
+# can we add mesh as an argument to this?
+nelements(solver::DG{<:RefElemData}, cache) = size(cache.u_values,2) 
```
Sure

jlchan replied:
I think similar changes might need to be made in the `analysis_callback` code. I haven't checked closely, will leave this conversation open until I can make those.

jlchan replied:
https://github.com/trixi-framework/Trixi.jl/blob/bdacf612e817881e12f1a4a80da465df57ab8f82/src/callbacks_step/analysis.jl#L191 

I'm hesitant to change this line b/c it seems like it might break things for other solvers.

ranocha replied:
I guess that's just a lesson that we learned from your PR: Methods like `nelements` etc. should also get the `mesh` as additional argument everywhere. Thus, we would use `nelements(mesh, solver, cache)`, `eachelement(mesh, solver, cache)` etc.

jlchan replied:
Should we make that a separate PR?

ranocha replied:
I'm fine with doing it here to verify that it's also working for your new solvers.

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+# !!! warning "Experimental features"
```
Could you please add this warning also to the example elixirs so that we can change things later without introducing breaking changes?

jlchan replied:
Done. Also tagged the other source files with the same warning.

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+# this is necessary for pretty printing
+Base.real(rd::RefElemData{Dims, Elem, ApproxType, Nfaces, RealT}) where {Dims, Elem, ApproxType, Nfaces, RealT} = RealT
+
+function Trixi.ndofs(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache)
```
```suggestion
function ndofs(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache)
```

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+  rd = dg.basis
+  return rd.Np * mesh.md.num_elements
+end
+Trixi.wrap_array(u_ode::StructArray, semi::Trixi.AbstractSemidiscretization) = u_ode
+Trixi.wrap_array(u_ode::StructArray, mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache) = u_ode
```
```suggestion
wrap_array(u_ode::StructArray, semi::AbstractSemidiscretization) = u_ode
wrap_array(u_ode::StructArray, mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache) = u_ode
```
Is the firth on really necessary? It should be redirected to the second method, I think?

jlchan replied:
Removed the first one; seems OK.

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+Trixi.wrap_array(u_ode::StructArray, mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache) = u_ode
+
+# interface with semidiscretization_hyperbolic
+function Trixi.digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, 
```
```suggestion
function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, 
```

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+  return StructArray{SVector{nvars, real(dg)}}(ntuple(_->similar(md.x),nvars))
+end
+
+function compute_coefficients!(u::StructArray, initial_condition, t,
+                 mesh::AbstractMeshData{Dim}, equations, dg::DG{<:RefElemData{Dim}}, cache) where {Dim}
```
```suggestion
function compute_coefficients!(u::StructArray, initial_condition, t,
                               mesh::AbstractMeshData{Dim}, equations, 
                               dg::DG{<:RefElemData{Dim}}, cache) where {Dim}
```

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+
+  num_quad_nodes = length(u_values) 
+  for i in Base.OneTo(num_quad_nodes) 
+    xyz_i = SVector{Dim}(getindex.(md.xyzq, i))
```
Is this type stable and efficient?

jlchan replied:
Yes, unless `@code_warntype` is lying to me (a la https://github.com/JuliaLang/julia/issues/32834).

jlchan replied:
`getindex.(md.xyzq,i)` is non-allocating since `md.xyzq::NTuple{Dim,Matrix{Tv}} where {Tv}`. It returns a tuple of coordinates, and I think converting to `SVector{Dim}` should be efficient.

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+end
+
+# interpolates from solution coefficients to face quadrature points
+function prolong2interfaces!(cache, u, mesh::AbstractMeshData, equations, 
+               surface_integral, dg::DG{<:RefElemData})
```
```suggestion
function prolong2interfaces!(cache, u, mesh::AbstractMeshData, equations, 
                             surface_integral, dg::DG{<:RefElemData})
```

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+  @unpack u_face_values, flux_face_values = cache 
+
+  num_face_nodes = length(u_face_values)
+  for i in Base.OneTo(num_face_nodes)
```
```suggestion
  for face in Base.OneTo(num_face_nodes)
```
We often try to use a bit more suggestive names for loop indices like this, e.g.
- `for element in eachelement(...)`
- `for boundary in achboundary(...)`
- `for interface in eachinterface(...)`
- `for mortar in eachmortar(...)`

It could be nice to use some similar style for these new solvers

jlchan replied:
Can do. I'll go ahead and use `mesh`, `equations`, `dg`, `cache` as per the discussion in https://github.com/trixi-framework/Trixi.jl/issues/661#issuecomment-866914977.

jlchan replied:
Actually, I think I'll leave off `equations` from the signature. I feel like that doesn't make much sense to include.

ranocha commented on src/solvers/dg_triangles/dg.jl:
```diff
@@ -0,0 +1,287 @@
+  for e in Base.OneTo(md.num_elements)
+    u_e = view(u_values, :, e) # u_values should already be computed from volume kernel
+
+    # todo: why do these lines allocate?
+    for i in Base.OneTo(rd.Nq)
+      flux_values[i] = source_terms(u_e[i], getindex.(md.xyzq, i, e), t, equations) 
+    end
```
What is returned by `getindex.(md.xyzq, i, e)`?

jlchan replied:
`md.xyzq` is a `Tuple{Matrix{Tv}} where {Tv}` so it returns a tuple consisting of `xq[i,e], yq[i,e]`, where `xq, yq` are coordinates of quadrature points.

jlchan replied:
I had tried converting it to an `SVector` and inlining `source_terms`, but they didn't seem to make much difference.

ranocha replied:
I will have a look at this part soon. How did you notice that it allocates (which elixir did you run)?

jlchan replied:
Thanks! I think any of the Euler elixirs should reproduce this, but I noticed this from the `@trixi_timeit` output for `examples/triangular_mesh_2D/elixir_euler_periodic_triangular_mesh.jl`.

ranocha replied:
`rhs!` is not specialized on `source_terms`. Thus, there is dynamic dispatch when calling `calc_sources!`. You can avoid that by forcing specialization of `rhs!` on `source_terms` as in 
```julia
function rhs!(du, u, t, mesh, equations,
              initial_condition, boundary_conditions, source_terms::SourceTerms,
              dg::DG{<:RefElemData}, cache) where {SourceTerms}
```
However, that should only be a minor performance improvement (if significant at all), since this dynamic dispatch is only happening once per call of `rhs!`.

ranocha commented on src/solvers/dg_triangles/mesh.jl:
```diff
@@ -0,0 +1,37 @@
+
+abstract type AbstractMeshData{Dim} end
+
+"""
+VertexMappedMesh describes a mesh which is constructed by an reference-to-physical 
```
```suggestion
"""
    VertexMappedMesh

`VertexMappedMesh` describes a mesh which is constructed by an reference-to-physical 
```

ranocha commented on src/solvers/dg_triangles/mesh.jl:
```diff
@@ -0,0 +1,37 @@
+  boundary_faces::Dict{Symbol,Vector{Int}}
+end
+
+Trixi.ndims(::VertexMappedMesh{Dim}) where {Dim} = Dim
```
```suggestion
Base.ndims(::VertexMappedMesh{Dim}) where {Dim} = Dim
```

ranocha commented on src/solvers/dg_triangles/mesh.jl:
```diff
@@ -0,0 +1,37 @@
+Trixi.ndims(::VertexMappedMesh{Dim}) where {Dim} = Dim
+
+"""
+  function VertexMappedMesh(VXYZ::NTuple{Dim,Vector{Tv}},EToV,rd::RefElemData;
+                            is_on_boundary=nothing,
+                            is_periodic::NTuple{Dim,Bool}=ntuple(_->false,Dim)) where {Dim,Tv}
+
+- `VXYZ` is a tuple of vectors of vertex coordinates
+- `EToV` is a matrix containing element-to-vertex connectivities for each element
+- `is_on_boundary` specifies boundary using Dict{Symbol,<:Function}
+- `is_periodic` is a tuple of booleans specifying periodicity = true/false in the (x,y,z) direction.
+"""
```
You can remove the `function` part from the docstring and just state the signature

ranocha commented on src/solvers/dg_triangles/mesh.jl:
```diff
@@ -0,0 +1,37 @@
+
+- `VXYZ` is a tuple of vectors of vertex coordinates
+- `EToV` is a matrix containing element-to-vertex connectivities for each element
+- `is_on_boundary` specifies boundary using Dict{Symbol,<:Function}
```
```suggestion
- `is_on_boundary` specifies boundary using a `Dict{Symbol,<:Function}`
```

ranocha commented on src/solvers/dg_triangles/mesh.jl:
```diff
@@ -0,0 +1,37 @@
+- `VXYZ` is a tuple of vectors of vertex coordinates
+- `EToV` is a matrix containing element-to-vertex connectivities for each element
+- `is_on_boundary` specifies boundary using Dict{Symbol,<:Function}
+- `is_periodic` is a tuple of booleans specifying periodicity = true/false in the (x,y,z) direction.
```
```suggestion
- `is_periodic` is a tuple of booleans specifying periodicity = `true`/`false` in the (x,y,z) direction.
```

ranocha commented on test/Project.toml:
```diff
@@ -8,6 +8,8 @@ MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195"
 OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
 Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+StartUpDG = "472ebc20-7c99-4d4b-9470-8fde4e9faa0f"
+StructArrays = "09ab397b-f2b6-538f-b94a-2f83cf4a842a"
```
You should add some `[compat]` bounds

jlchan replied:
Done. The compat bounds for StructArrays are set to 0.6, which should be the next tagged release.

ranocha commented on test/runtests.jl:
```diff
@@ -50,6 +50,10 @@ const TRIXI_NTHREADS   = clamp(Sys.CPU_THREADS, 2, 3)
     include("test_examples_2d_part3.jl")
   end
 
+  @time if TRIXI_TEST == "all" || TRIXI_TEST == "2d_tri"
```
```suggestion
  @time if TRIXI_TEST == "all" || TRIXI_TEST == "2d_triangles"
```

ranocha commented on test/test_examples_2d_triangular_mesh.jl:
```diff
@@ -0,0 +1,35 @@
+module TestExamples2DTri
+
+using Test
+# using StructArrays
+using StartUpDG
```
```suggestion
```
Those are already included in the elixirs, aren't they?

ranocha commented on test/test_examples_2d_triangular_mesh.jl:
```diff
@@ -0,0 +1,35 @@
+    
+    @test true # dummy test
+
+    # @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_triangular_mesh.jl"),
+    #   l2 = [6.27122850947113e-11, 6.791612221354976e-11, 6.791612221249498e-11, 6.466533604304294e-10], 
+    #   linf = [1.8523039727380564e-5, 1.6923451381156696e-5, 1.71384595346602e-5, 4.586131120554171e-5]
+    # )
+
+    # @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_periodic_triangular_mesh.jl"),
+    #   l2 = [6.943606879787733e-11, 7.474517827181276e-11, 7.474517827154942e-11, 6.637284410683966e-10], 
+    #   linf = [1.8302801320757567e-5, 1.6548179149200593e-5, 1.6886330820753415e-5, 4.5056371695828545e-5]
+    # )
+
+    # @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_ape_triangular_mesh.jl"),
+    #   l2 = [4.293692357434845e-6, 4.293692357445338e-6, 2.069046541101411e-6, 3.978858062534657e-32, 3.978858062534657e-32, 6.3661729000554515e-31, 6.3661729000554515e-31], 
+    #   linf = [0.005093360833536398, 0.0010859409295087552, 0.00012810312361999365, 1.6653345369377348e-16, 1.6653345369377348e-16, 6.661338147750939e-16, 6.661338147750939e-16]
+    # )
```
Needs to be activated once upstream issues are fixed.

ranocha reviewed (changes requested):
Thanks a lot for working on this, @jlchan! Please find below some comments and questions.

- Did you run some convergence tests using the new setups?
- How does the performance look like compared to the other solvers we have in Trixi right now?
- Could you please add another example using a quad mesh based on your infrastructure?
- Could you please add some note to `NEWS.md` about this new (experimental) feature?

jlchan commented:
> Thanks a lot for working on this, @jlchan! Please find below some comments and questions.
> 
>     * Did you run some convergence tests using the new setups?

Not yet. Are there utilities for running convergence tests or an example I should look at?
> 
>     * How does the performance look like compared to the other solvers we have in Trixi right now?

Good question - I remember a PID of about 1.9e-7 for N=3 on triangles, but I need to do a more careful comparison.
> 
>     * Could you please add another example using a quad mesh based on your infrastructure?

Done - and added a triangular SBP example too. 
> 
>     * Could you please add some note to `NEWS.md` about this new (experimental) feature?

Done!

@sloede should I hold off from adding commits addressing @ranocha's comments until after you've had a chance to review?

ranocha commented:
> @sloede should I hold off from adding commits addressing @ranocha's comments until after you've had a chance to review?

Thanks a lot! I will have another look at the changes later (tomorrow?). I don't think you should wait to push changes. Usually, @sloede and I would comment on some similar things and he has already quite a lot on his TODO list. Thus, you will save him time if you do everything you want right now and he can look at the combined changes later.

jlchan commented:
That sounds great. Changes pushed.

ranocha commented:
> Are there utilities for running convergence tests or an example I should look at?

We have a function `convergence_test` https://github.com/trixi-framework/Trixi.jl/blob/bdacf612e817881e12f1a4a80da465df57ab8f82/src/auxiliary/special_elixirs.jl#L43
You would need to specialize https://github.com/trixi-framework/Trixi.jl/blob/bdacf612e817881e12f1a4a80da465df57ab8f82/src/auxiliary/special_elixirs.jl#L178 and https://github.com/trixi-framework/Trixi.jl/blob/bdacf612e817881e12f1a4a80da465df57ab8f82/src/auxiliary/special_elixirs.jl#L215 such that it can handle your measure of the discretization resolution.

ranocha commented:
Minor note: If you go to the "Files changed" tab, you can add suggestions to a batch such that you can apply all of them together. If you apply every suggestion on it's own, we will get a lot of commits and CI will be triggered for every single commit, which basically means that we would have to wait several days until all CI runs are finished (since there is only a limited amount of parallel CI time for open source projects). I canceled some CI runs to prevent this a few minutes ago, so you might receive some emails from GitHub informing you about this.

ranocha commented on examples/triangular_mesh_2D/elixir_ape_triangular_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+            dt = .5*dt0, save_everystep=false, callback=callbacks);
+summary_callback() # print the timer summary
+
+# u = sol.u[end]
```
```suggestion
```

ranocha commented on examples/triangular_mesh_2D/elixir_euler_quadrilateral_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+            dt = .5*dt0, save_everystep=false, callback=callbacks);
+summary_callback() # print the timer summary
+
+# u = sol.u[end]
```
```suggestion
```

ranocha commented on examples/triangular_mesh_2D/elixir_euler_sbp_triangular_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+            dt = .5*dt0, save_everystep=false, callback=callbacks);
+summary_callback() # print the timer summary
+
+# u = sol.u[end]
```
```suggestion
```

ranocha commented on examples/triangular_mesh_2D/elixir_euler_sbp_triangular_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+using StartUpDG, StructArrays
+using Trixi, OrdinaryDiffEq
+
+rd = RefElemData(Tri(),SBP(), N=4)
```
```suggestion
rd = RefElemData(Tri(), SBP(), N=4)
```
This one uses the SBP operators on simplices of Hicken et al., doesn't it?

jlchan replied:
It depends on the options. Currently, it defaults to a set of degree SBP operators built from nodes with degree `2p-1` quadrature accuracy and `p+2` Lobatto nodes on the faces (from Ethan Kubatko). One can access Hicken's nodes via
```julia
rd = RefElemData(Tri(),SBP(), N, quadrature_strength=2*N) # Hicken nodes
rd = RefElemData(Tri(), SBP(), N, quadrature_strength=2*N-1, quad_rule_face=:Legendre) # Chen-Shu nodes
```

The kwarg API should probably be improved - I'm thinking about adding these as symbol fields in the `SBP()` type instead.

ranocha replied:
Thanks, that sounds good :+1:

ranocha commented on examples/triangular_mesh_2D/elixir_euler_triangular_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+            dt = .5*dt0, save_everystep=false, callback=callbacks);
+summary_callback() # print the timer summary
+
+# u = sol.u[end]
```
```suggestion
```

jlchan commented:
> Minor note: If you go to the "Files changed" tab, you can add suggestions to a batch such that you can apply all of them together. If you apply every suggestion on it's own, we will get a lot of commits and CI will be triggered for every single commit, which basically means that we would have to wait several days until all CI runs are finished (since there is only a limited amount of parallel CI time for open source projects). I canceled some CI runs to prevent this a few minutes ago, so you might receive some emails from GitHub informing you about this.

I saw the slew of emails - thank you for cancelling the runs and for the heads up!

jlchan commented:
Can I still add `[ci skip]` to commit messages to skip CI in Trixi?

ranocha commented:
> Can I still add `[ci skip]` to commit messages to skip CI in Trixi?
> 
> Trying it out.

It's `[skip ci]` nowadays

jlchan commented:
> You would need to specialize
> 
> https://github.com/trixi-framework/Trixi.jl/blob/bdacf612e817881e12f1a4a80da465df57ab8f82/src/auxiliary/special_elixirs.jl#L178
> and
> 
> https://github.com/trixi-framework/Trixi.jl/blob/bdacf612e817881e12f1a4a80da465df57ab8f82/src/auxiliary/special_elixirs.jl#L215
> such that it can handle your measure of the discretization resolution.

I just ended up using those functions as-is with `cells_per_dimension` as my resolution. Convergence test works, and we got optimal `O(h^4)` rates of convergence for `N=3` for both L2, Linf norms. Will commit soon (also caught a mistake in naming).

jlchan commented:
For `polydeg=3` or `N=3`, the elixirs `elixir_euler_triangular_mesh_convergence.jl` and `elixir_euler_unstructured_quad_basic.jl` both give a PID of around 2e-7 on average. I haven't used threading or other optimizations to the triangular mesh solvers yet though.

ranocha commented:
> I just ended up using those functions as-is with `cells_per_dimension` as my resolution. Convergence test works, and we got optimal `O(h^4)` rates of convergence for `N=3` for both L2, Linf norms. With an initial resolution of `8x8` cells, we get
> 
> ```julia
>   :variables => ("rho", "rho_v1", "rho_v2", "rho_e")
>   :l2        => [4.12765, 4.07351, 4.07351, 4.12278]
>   :linf      => [3.95636, 3.93671, 4.19283, 4.25113]
> ```
> 
> I like the usage of the `analysis_callback` as a common interface for collecting information.

Cool :sunglasses: 

> It would also be nice to have a way to specify the symbol for resolution, e.g., to allow for other names than `:cells_per_dimension`. Would this be difficult?

It's a bit more complicated than I would like since we need to operate at the level of expressions (basically via the 2-argument version of `include`). I'm not sure whether there is a really simple and extensible way to do so, but we have already hard-coded the possible keywords `initial_refinement_level` (`TreeMesh`) and `cells_per_dimension` (`CurvedMesh`, which shall become `StructuredMesh`). We could as well add a third possible keyword...

ranocha commented:
> For `polydeg=3` or `N=3`, the elixirs `elixir_euler_triangular_mesh_convergence.jl` and `elixir_euler_unstructured_quad_basic.jl` both give a PID of around 2e-7 on average. I haven't used threading or other optimizations to the triangular mesh solvers yet though.

Nice :+1: Since you're using the physical and numerical fluxes from Trixi, these new solvers will also benefit from the performance improvements of #643 (standard DG is also improved although EC is in the title of that PR).

ranocha commented on test/Project.toml:
```diff
@@ -19,5 +19,5 @@ ForwardDiff = "0.10"
 OrdinaryDiffEq = "5.44 - 5.50, 5.51.1"
 Plots = "1.9"
-StartUpDG = "0.8"
+StartUpDG = "0.8, 0.9.1, 0.10"
```
```suggestion
StartUpDG = "0.10"
```
Some of your commit messages read like "We should use the new version of StartUpDG". Is that correct? If so, I would propose to just request the newest version here.

jlchan replied:
Sounds good. 

The flood of versions came b/c I broke the public API on accident...

ranocha commented:
It would also be really great if you could describe some of the conventions (such as `rd.wf, rd.wq` are weights at the face/volume quadrature nodes, choice of coordinate names in terms like `rxJ, sxJ, ryJ, syJ`, meaning of `rd.Vq` etc.) you use in the docs. You could create a new file in https://github.com/trixi-framework/Trixi.jl/tree/main/docs/src/meshes and include it in the build process of the docs at https://github.com/trixi-framework/Trixi.jl/blob/main/docs/make.jl#L47-L50.

jlchan commented:
> It's a bit more complicated than I would like since we need to operate at the level of expressions (basically via the 2-argument version of `include`). I'm not sure whether there is a really simple and extensible way to do so, but we have already hard-coded the possible keywords `initial_refinement_level` (`TreeMesh`) and `cells_per_dimension` (`CurvedMesh`, which shall become `StructuredMesh`). We could as well add a third possible keyword...

No worries - I'm not familiar with macros, so I wasn't sure how difficult this would be. It works great as-is anyways.

> It would also be really great if you could describe some of the conventions (such as `rd.wf, rd.wq` are weights at the face/volume quadrature nodes, choice of coordinate names in terms like `rxJ, sxJ, ryJ, syJ`, meaning of `rd.Vq` etc.) you use in the docs. You could create a new file in https://github.com/trixi-framework/Trixi.jl/tree/main/docs/src/meshes and include it in the build process of the docs at https://github.com/trixi-framework/Trixi.jl/blob/main/docs/make.jl#L47-L50.

Sure, I can do that. I'm planning on writing an overview of what I use within Trixi and then referring users to the StartUpDG docs. Does that sound reasonable?

ranocha commented:
> Sure, I can do that. I'm planning on writing an overview of what I use within Trixi and then referring users to the StartUpDG docs. Does that sound reasonable?

Perfect :+1:

ranocha commented on docs/src/meshes/MeshDataMeshes.md:
```diff
@@ -0,0 +1,70 @@
+* `md.xyz` is a tuple of matrices `md.x`, `md.y`, `md.z`, where column `e` contains coordinates of physical interpolation points. 
+* `md.xyzq` is a tuple of matrices `md.xq`, `md.yq`, `md.zq`, where column `e` contains coordinates of physical quadrature points. 
+* `md.rxJ, md.sxJ, ...` are matrices where column `e` contains values of ``J\frac{\partial r}{\partial x}``, ``J\frac{\partial s}{\partial x}``, etc. at nodal interpolation points on the element `e`.
+* `md.J` is a matrix where column `e` contains values of ``J`` at nodal interpolation points.
```
```suggestion
* `md.J` is a matrix where column `e` contains values of the Jacobian ``J`` at nodal interpolation points.
```

ranocha replied:
`J` is the Jacobian, isn't it?

jlchan replied:
Yes, that's correct

ranocha commented on docs/src/meshes/MeshDataMeshes.md:
```diff
@@ -0,0 +1,70 @@
+* `md.xyzq` is a tuple of matrices `md.xq`, `md.yq`, `md.zq`, where column `e` contains coordinates of physical quadrature points. 
+* `md.rxJ, md.sxJ, ...` are matrices where column `e` contains values of ``J\frac{\partial r}{\partial x}``, ``J\frac{\partial s}{\partial x}``, etc. at nodal interpolation points on the element `e`.
+* `md.J` is a matrix where column `e` contains values of ``J`` at nodal interpolation points.
+* `md.sJ` is a matrix where column `e` contains values of the face Jacobian (e.g., determinant of the geometric mapping between a physical face and a reference face) at face quadrature points.
```
What does `s` mean here? Based on the description so far, I would have expected an `f` for "face"... What am I missing?

jlchan replied:
This unfortunately is an inconsistency on my part. The original Matlab codes used 'sJ' and we just stuck with it. 

I suggest that I switch to 'md.Jf'  in StartUpDg and use that within Trixi. I can overload 'md.sJ' as a private property for backwards compatibility.

ranocha replied:
Sounds good, thanks :+1:

jlchan replied:
Fixed in the updated docs.

ranocha commented on docs/src/meshes/MeshDataMeshes.md:
```diff
@@ -0,0 +1,70 @@
+For more details, please see the [StartUpDG.jl docs](https://jlchan.github.io/StartUpDG.jl/dev/). 
+## Special options
+
+Trixi solvers on simplicial meshes use `DG` solver types with the basis field set as a `RefElemData` type. By default, `RefElemData(Tri(), N)` will generate data for a degree ``N`` polynomial approximation on a triangle. There are also several parameters which can be tweaked:
```
```suggestion
Trixi solvers on simplicial meshes use `DG` solver types with the `basis` field set as a `RefElemData` type. By default, `RefElemData(Tri(), N)` will generate data for a degree ``N`` polynomial approximation on a triangle. There are also several parameters which can be tweaked:
```

How strong do you prefer to use `N` as variable name for the polynomial degree? It might make sense to use `polydeg` (here and in the elixirs) as we do in other parts of Trixi. 
Background: We had quite a discussion about this in the past since some core developers grew up using `N` for the polynomial degree, too. In the end, we decided to use `polydeg` since that makes it easier to parse for people from other schools. 
Anyways, we need to find a balance between Trixi and StartUpDG, so I would not insist on a specific variable name here.

jlchan replied:
I'm happy to use polydeg within any Trixi elixirs. 

I'm less sure about changing 'N' in 'RefElemData' though. What do you think about overloading 'getproperty' so that 'rd.polydeg' returns N, but not making that property public?

ranocha replied:
Sounds like an option. On the other hand, I think it's fine to query the polynomial degree using the function `polydeg` inside Trixi (if necessary at all). I was more thinking about the argument name used to set up everything.

jlchan replied:
So 'RefElemData(Tri(), polydeg)' in the elixir setup? That sounds great to me

ranocha commented on docs/src/meshes/MeshDataMeshes.md:
```diff
@@ -0,0 +1,70 @@
+
+Trixi solvers on simplicial meshes use `DG` solver types with the basis field set as a `RefElemData` type. By default, `RefElemData(Tri(), N)` will generate data for a degree ``N`` polynomial approximation on a triangle. There are also several parameters which can be tweaked:
+
+* `RefElemData(Tri(), N, quad_rule_vol = quad_nodes(Tri(), Nq))` will substitute in a volume quadrature rule of degree `Nq` instead of the default (which is a quadrature rule of degree `N`). Here, a degree `Nq` rule will be exact for at least degree `2*Nq` integrands (such that the mass matrix is integrated exactly).
```
Can you specify a quadrature that is exact for polynomials of odd degree (such as Lobatto-Legendre quadrature)? Would that make sense at all?

jlchan replied:
Yes, 'quad_nodes' just calls another quadrature routine with argument '2*Nq'. I can make a note of that in the docs. 

I defaulted to 'quad_nodes' returning a quadrature rule of degree '2*Nq' to make the behavior in 1d ('quad_nodes(N)' returns an N+1 Gauss Legendre rule exact for at least a degree 2N integrand) consistent with 2D.

ranocha replied:
Sounds good 👍An additional note would be perfect.

ranocha commented on docs/src/meshes/MeshDataMeshes.md:
```diff
@@ -0,0 +1,70 @@
+Trixi solvers on simplicial meshes use `DG` solver types with the basis field set as a `RefElemData` type. By default, `RefElemData(Tri(), N)` will generate data for a degree ``N`` polynomial approximation on a triangle. There are also several parameters which can be tweaked:
+
+* `RefElemData(Tri(), N, quad_rule_vol = quad_nodes(Tri(), Nq))` will substitute in a volume quadrature rule of degree `Nq` instead of the default (which is a quadrature rule of degree `N`). Here, a degree `Nq` rule will be exact for at least degree `2*Nq` integrands (such that the mass matrix is integrated exactly).
+* `RefElemData(Tri(), N, quad_rule_face = quad_nodes(Line(), Nq))` will use a face quadrature rule of degree `Nq` rather than the default. 
```
This means that `quad_rule_face` integrates polynomials of degree `2*Nq` exactly, doesn't it?

jlchan replied:
Yep!

ranocha commented on docs/src/meshes/MeshDataMeshes.md:
```diff
@@ -0,0 +1,70 @@
+
+* `RefElemData(Tri(), N, quad_rule_vol = quad_nodes(Tri(), Nq))` will substitute in a volume quadrature rule of degree `Nq` instead of the default (which is a quadrature rule of degree `N`). Here, a degree `Nq` rule will be exact for at least degree `2*Nq` integrands (such that the mass matrix is integrated exactly).
+* `RefElemData(Tri(), N, quad_rule_face = quad_nodes(Line(), Nq))` will use a face quadrature rule of degree `Nq` rather than the default. 
+* `RefElemData(Tri(), SBP(), N)` will generate a reference element corresponding to a multi-dimensional SBP discretization. Types of SBP discretizations available include `SBP{Kubatko{LobattoFaceNodes}}()` (the default choice), `SBP{Kubatko{LegendreFaceNodes}}()`, and `SBP{Hicken}()`. Descriptions of each can be found in the [StartUpDG.jl docs](https://jlchan.github.io/StartUpDG.jl/dev/RefElemDat ...
```
Is there a paper for the methods of Kubatko? I could only find paper references for the other two options at https://jlchan.github.io/StartUpDG.jl/dev/RefElemData/#RefElemData-based-on-SBP-finite-differences.

jlchan replied:
Unfortunately no - he just sent the nodes over private communication. The 'Kubatko{LegendreFaceNodes}' correspond to the SBP nodes of Chen and Shu 2017 though. Do you think we should cite that for the Kubatko nodes, with a note that the Kubatko rules are more general?

ranocha replied:
That's a good idea :+1:

ranocha commented on src/solvers/dg_simplices/mesh.jl:
```diff
@@ -0,0 +1,57 @@
+Base.ndims(::VertexMappedMesh{Dim}) where {Dim} = Dim
+
+"""
+  VertexMappedMesh(VXYZ::NTuple{Dim, Vector{Tv}}, EToV, rd::RefElemData;
+                   is_on_boundary = nothing,
+                   is_periodic::NTuple{Dim, Bool} = ntuple(_->false, Dim)) where {Dim, Tv}
```
```suggestion
    VertexMappedMesh(VXYZ::NTuple{Dim, Vector{Tv}}, EToV, rd::RefElemData;
                     is_on_boundary = nothing,
                     is_periodic::NTuple{Dim, Bool} = ntuple(_->false, Dim)) where {Dim, Tv}
```
I think you need four spaces to mark this as code

ranocha commented on src/solvers/dg_simplices/mesh.jl:
```diff
@@ -0,0 +1,57 @@
+end
+
+"""
+  VertexMappedMesh(VXYZ::NTuple{Dim, Vector{Tv}}, EToV, rd::RefElemData;
+                   triangulateIO, boundary_dict::Dict{Symbol, Int})
```
```suggestion
    VertexMappedMesh(VXYZ::NTuple{Dim, Vector{Tv}}, EToV, rd::RefElemData;
                     triangulateIO, boundary_dict::Dict{Symbol, Int})
```

ranocha reviewed (commented):
Nice documentation, @jlchan :+1: Thanks a lot for this!

sloede commented on docs/src/meshes/MeshDataMeshes.md:
```diff
@@ -0,0 +1,70 @@
+# Unstructured simplicial meshes
+
+Support for simplicial meshes uses the [StartUpDG.jl](https://github.com/jlchan/StartUpDG.jl) package. These meshes are constructed by specifying a list of vertex coordinates `VX, VY, VZ` and a connectivity matrix `EToV` where `EToV[e,:]` gives the vertices which correspond to element `e`. 
```
Not bikeshedding IMHO! At least with GitHub it makes looking at diffs much easier if the lines fit well within the browser window.

jlchan replied:
No problem, and thanks for the feedback!

sloede replied:
Isn't it 3:30am in Houston right now? 🤔

jlchan replied:
Had trouble sleeping 😅

jlchan commented on docs/src/meshes/MeshDataMeshes.md:
```diff
@@ -0,0 +1,70 @@
+* variables such as `r, s,...` and `x, y,...` correspond to values at nodal interpolation points. 
+* variables ending in `q` (e.g., `rq, sq,...` and `xq, yq,...`) correspond to values at volume quadrature points. 
+* variables ending in `f` (e.g., `rf, sf,...` and `xf, yf,...`) correspond to values at face quadrature points. 
+* variables ending in `p` (e.g., `rp, sp,...`)
```
Oops, I forgot to finish that sentence. Will add

jlchan commented on src/semidiscretization/semidiscretization_hyperbolic.jl:
```diff
@@ -83,15 +83,15 @@ end
 digest_boundary_conditions(boundary_conditions, mesh, solver, cache) = boundary_conditions
 
-function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, mesh, solver, cache) where {Keys, ValueTypes<:NTuple{2,Any}} # 1D
+function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, mesh::AbstractMesh{1}, solver, cache) where {Keys, ValueTypes<:NTuple{2,Any}} # 1D
```
Can these be fully specialized to `TreeMesh{Dim}`?

ranocha replied:
Should work, I think :+1:

jlchan commented:
Thanks for the reviews @ranocha! I think I fixed most of the issues you caught.

ranocha commented on src/solvers/dg_simplices/dg.jl:
```diff
@@ -0,0 +1,294 @@
+  rd = dg.basis
+  @unpack u_values = cache
+
+  for i in eachquadnodetotal(mesh,dg,cache)
```
I have to admit that I have some trouble reading something like `eachquadnodetotal`, where four words are squashed together... Something like `eachnode` with only two words is fine for me, but I would prefer adding underscores or trying to find shorter names for `eachquadnodetotal`, if possible.

jlchan replied:
I agree.

I can think of a few options:
1. Change the name: `each_quad_node_total` or `each_quad_node_global`
2. Add an outer loop over elements and use `eachnode` to iterate over nodes in an element.
3. We could use dispatch to distinguish between nodes on an element and nodes over the entire mesh (e.g., `each_quad_node(mesh, solver, cache)` for global nodes over the entire mesh, `each_quad_node(solver, cache)` for nodes over a single element.

I admit I don't like 2) since it would be easier to parallelize across all nodes with a single for loop. Do you have a preference between 1) or 3)

ranocha replied:
I would prefer 1) since it's more clear to me.

ranocha commented on src/solvers/dg_simplices/dg.jl:
```diff
@@ -0,0 +1,294 @@
+
+  # todo: factor common storage into a struct (MeshDataCache?) for reuse 
+  # storage for volume quadrature values, face quadrature values, flux values
+  u_values = StructArray{SVector{nvars, RealT}}(ntuple(_->zeros(rd.Nq, md.num_elements), nvars))
+  u_face_values = StructArray{SVector{nvars, RealT}}(ntuple(_->zeros(rd.Nfq, md.num_elements), nvars))
```
```suggestion
  u_values = StructArray{SVector{nvars, uEltype}}(ntuple(_->zeros(rd.Nq, md.num_elements), nvars))
  u_face_values = StructArray{SVector{nvars, uEltype}}(ntuple(_->zeros(rd.Nfq, md.num_elements), nvars))
```

ranocha commented on src/solvers/dg_simplices/dg.jl:
```diff
@@ -0,0 +1,294 @@
+  flux_face_values = similar(u_face_values)
+
+  # local storage for fluxes
+  flux_values = StructArray{SVector{nvars,RealT}}(ntuple(_->zeros(rd.Nq), nvars))
```
```suggestion
  flux_values = StructArray{SVector{nvars, uEltype}}(ntuple(_->zeros(rd.Nq), nvars))
```

ranocha reviewed (commented):
Thanks a lot, @jlchan. Do you think there is anything left to do in this PR except what's listed below?

- [ ] Add tests once a new release of StructArrays is published
- [ ] Fix `# todo` notes added here or convert them into `# TODO: Simplices` (or something like that to make it easier to find them) and create a new issue tracking them
- [ ] Fix remaining open review comments
- [ ] Get a review by @sloede

jlchan commented:
> Thanks a lot, @jlchan. Do you think there is anything left to do in this PR except what's listed below?
>     * [ ]  Add tests once a new release of StructArrays is published
>     * [ ]  Fix `# todo` notes added here or convert them into `# TODO: Simplices` (or something like that to make it easier to find them) and create a new issue tracking them
>     * [ ]  Fix remaining open review comments
>     * [ ]  Get a review by @sloede

Not for this PR, I think. For each of these points:
1. Will have those tests re-activated soon.
2. Done. 
3. https://github.com/trixi-framework/Trixi.jl/issues/675 
4. Since I imagine @sloede seems pretty busy with Juliacon prep these few weeks, I'll just leave this PR as-is until he can take a look at it. I'll branch from this PR so that I can work on other stuff for the todo list. Does that sound OK?

ranocha commented:
> 4\. I'll branch from this PR so that I can work on other stuff for the todo list. Does that sound OK?

Sounds good to me, thanks!

ranocha commented on Project.toml:
```diff
@@ -13,6 +13,7 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 LinearMaps = "7a12625a-238d-50fd-b39a-03d52299707e"
 LoopVectorization = "bdcacae8-1622-11e9-2a5c-532679323890"
 MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195"
+Octavian = "6fd5a793-0b7e-452c-907f-f8bfe9c57db4"
```
Should get a `[compat]` bound

jlchan replied:
Yeah, sorry - was waiting to see if the CI tests passed (they did not). Will add compat bound soon

ranocha commented on src/solvers/dg_simplices/analysis.jl:
```diff
@@ -9,19 +8,17 @@ function calc_error_norms(func, u::StructArray, t, analyzer,
+  StructArrays.foreachfield(mul_by!(rd.Vq), u_values, u)   
 
-  return component_l2_errors, component_linf_errors
+  component_l2_errors = zero(eltype(u_values))
+  component_linf_errors = similar(component_l2_errors)
```
When using StaticArrays, `component_l2_errors` will be an `SVector` but `component_linf_errors` will be an `MVector - is that the desired behavior?

jlchan replied:
That is not intended - I didn't realize that it would return `MVector`. Thanks for catching it!

ranocha commented on test/Project.toml:
```diff
@@ -5,17 +5,17 @@ ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 MPI = "0.16, 0.17, 0.18"
-OrdinaryDiffEq = "5.44 - 5.50, 5.51.1"
-Plots = "1.9"
+OrdinaryDiffEq = "5.44 - 5.50, 5.51.1, 5.59"
```
Why did you change the `[compat]` entry for OrdinaryDiffEq?

jlchan replied:
I was using OrdinaryDiffEq 5.59 for local testing, and wanted to check if this worked for CI tests as well. This was just an experiment, and I can revert back to the original compat.

ranocha commented on test/Project.toml:
```diff
@@ -5,17 +5,17 @@ ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 ForwardDiff = "0.10"
 MPI = "0.16, 0.17, 0.18"
-OrdinaryDiffEq = "5.44 - 5.50, 5.51.1"
-Plots = "1.9"
```
Need to remember to add this again after upstream issues are resolved

ranocha commented on test/test_examples_2d_triangular_mesh.jl:
```diff
@@ -8,26 +8,34 @@ include("test_trixi.jl")
-@testset "Triangular mesh tests" begin
-  @trixi_testset "Euler equations" begin
+@testset "Triangular mesh tests" begin    
+    # @test true # dummy test
```
```suggestion
```

ranocha commented:
What do you think about the following proposal, @jlchan ? We merge this PR into `dev`. Then, you can make more PRs to `dev` that I will review, for example adding flux differencing or other entropy-based methods. These PRs will be not too large so that it's easier for me to review them. When you're satisfied with the setup of your solvers in Trixi, we can merge `dev` into `main`. @sloede can review `dev` when he finds some time in the future.
We used this approach for other large changes successfully in the past, e.g., for the `P4estMesh`.

jlchan commented:
> What do you think about the following proposal, @jlchan ? We merge this PR into `dev`. Then, you can make more PRs to `dev` that I will review, for example adding flux differencing or other entropy-based methods. These PRs will be not too large so that it's easier for me to review them. When you're satisfied with the setup of your solvers in Trixi, we can merge `dev` into `main`. @sloede can review `dev` when he finds some time in the future.
> We used this approach for other large changes successfully in the past, e.g., for the `P4estMesh`.

That sounds great to me. Two questions

1. how is `dev` managed? Right now, I just merge `main` into my branch whenever it's updated. Can I do the same with `dev` or does it need one of the core devs to merge updates from main into it?
2. at the moment, Plots tests still aren't supported (will need Plots devs to address until https://github.com/JuliaPlots/Plots.jl/pull/3588). Is it OK to merge to dev without those tests active?

ranocha commented:
I guess you will work mostly on one PR at the same time. Thus, you can merge `main` into the branch of your currently active PR. If you really want to merge `main` into `dev` directly, just ping me.

ranocha commented:
2. Yes. `dev` is some kind of WIP area, so that's okay for me

jlchan commented:
Great! Then feel free to merge into `dev` whenever.

jlchan commented:
Actually - let me add a few final commits (mainly removing dimensional dependence from a few places).

ranocha commented:
Sure - ping me when you're ready (and remove the WIP from the title of the PR). I will take care of the rest

jlchan commented:
@ranocha should be ready for merging. Went ahead and added tets too 😁

ranocha commented on src/equations/compressible_euler_3d.jl:
```diff
@@ -791,6 +791,31 @@ end
 
+@inline function min_max_speed_naive(u_ll, u_rr, normal_direction::AbstractVector,
+                                     equations::CompressibleEulerEquations3D)
+  rho_ll, rho_v1_ll, rho_v2_ll, rho_v3_ll, rho_e_ll = u_ll
+  rho_rr, rho_v1_rr, rho_v2_rr, rho_v3_rr,  rho_e_rr = u_rr
+
+  # Calculate primitive variables and speed of sound
+  v1_ll = rho_v1_ll / rho_ll
+  v2_ll = rho_v2_ll / rho_ll
+  v3_ll = rho_v3_ll / rho_ll  
+  p_ll = (equations.gamma - 1) * (rho_e_ll - 1/2 * rho_ll * (v1_ll^2 + v2_ll^2 + v3_ll^2))
+
+  v1_rr = rho_v1_rr / rho_rr
+  v2_rr = rho_v2_rr / rho_rr
+  v3_rr = rho_v3_rr / rho_rr  
+  p_rr = (equations.gamma - 1) * (rho_e_rr - 1/2 * rho_rr * (v1_rr^2 + v2_rr^2 + v3_rr^2))
```
```suggestion
  rho_ll, v1_ll, v2_ll, v3_ll, p_ll = cons2prim(u_ll, equations)
  rho_rr, v1_rr, v2_rr, v3_rr, p_rr = cons2prim(u_rr, equations)
```
Preparation for #686

jlchan replied:
Thanks for adding that for me!

ranocha commented on examples/simplicial_mesh/elixir_euler_tet_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+source_terms = source_terms_convergence_test
+
+# example where we tag two separate boundary segments of the mesh
+top_boundary(x,y,z,tol=50*eps()) = abs(y-1)<tol 
+rest_of_boundary(x,y,z,tol=50*eps()) = !top_boundary(x,y,z,tol)
```
```suggestion
top_boundary(x, y, z, tol=50*eps()) = abs(y - 1) < tol 
rest_of_boundary(x, y, z, tol=50*eps()) = !top_boundary(x, y, z, tol)
```

ranocha commented on src/solvers/dg_simplices/dg.jl:
```diff
@@ -1,14 +1,17 @@
-# Don't use `matmul!` for the following 2 functions: 5-arg `matmul!` hangs when applied to vector 
-# arguments. See https://github.com/JuliaLinearAlgebra/Octavian.jl/issues/103. 
+# specialize for SBP operators since matmul! doesn't work for `UniformScaling` types
+mul_by!(A::UniformScaling) = let A = A
+  @inline (out, x)->out .= x
+end
```
That's problematic:
```juliajulia> using LinearAlgebra

julia> typeof(I)
UniformScaling{Bool}

julia> UniformScaling(2.0)
UniformScaling{Float64}
2.0*I
```
It handles the situation where you pass `I` correctly but will be false in general... Maybe restrict it to `UniformScaling{Bool}` and hope that nobody uses
```julia
julia> UniformScaling(false)
UniformScaling{Bool}
false*I
```
Alternatively, you could just use `mul!` here, which might be slightly more costly due to an additional multiplication but would be correct in general.

jlchan replied:
Hm, good point. I prefer the safer `mul!` fallback for now. 

This can also be avoided by dispatching based on `DG` types. Evaluating the solution at quadrature points requires a matrix multiply, as does performing an L2 projection. For collocated discretizations like DGSEM or nodal SBP, I just set the relevant matrices to `I` as a way to "skip" matrix multiplication. 

How about we use the `mul!` fallback for now, and I add this to the todo-list https://github.com/trixi-framework/Trixi.jl/issues/675?

ranocha replied:
Sounds good :+1:

ranocha commented on src/solvers/dg_simplices/dg.jl:
```diff
@@ -27,6 +30,8 @@ const DGWeakForm{Dims, ElemType} = DG{<:RefElemData{Dims, ElemType}, Mortar,
 # this is necessary for pretty printing
 Base.real(rd::RefElemData{Dims, Elem, ApproxType, Nfaces, RealT}) where {Dims, Elem, ApproxType, Nfaces, RealT} = RealT
 
+eachdim(mesh::AbstractMeshData{Dim}, dg::DG{<:RefElemData{Dim}}, cache) where {Dim} = Base.OneTo(Dim)
```
Suggestion:
```suggestion
eachdim(mesh::AbstractMeshData{Dim}) where {Dim} = Base.OneTo(Dim)
```
Do you use `ndims(mesh::AbstractMeshData)`? Looks like it might be a good opportunity to define a generic `eachdim(whatever) = Base.OneTo(ndims(whatever))`...

jlchan replied:
I don't think I do. I may have thought `ndims(mesh)` was used by Trixi's `Base.show` for some reason. 

I'll define `eachdim` in terms of `ndims` then.

ranocha replied:
Yeah, we use it mostly for `show` - but it's available and fast, so we can use it here :wink:

ranocha commented on src/solvers/dg_simplices/dg.jl:
```diff
@@ -98,43 +104,43 @@ function create_cache(mesh::VertexMappedMesh, equations, dg::DG, RealT, uEltype)
-    StructArrays.foreachfield(mul_by_accum!(invMQsTrW, syJ[1,e]), view(du,:,e), flux_values)
+    
+    flux_values = flux_values_threaded[Threads.threadid()]
+    for i in eachdim(mesh, dg, cache) 
+      flux_values .= flux.(view(u_values,:,e), i, equations)
+      for j in eachdim(mesh, dg, cache)
+        StructArrays.foreachfield(mul_by_accum!(invMQrstTrW[j], rstxyzJ[i,j][1,e]), 
+                                  view(du,:,e), flux_values)
+      end
+    end
```
Would need to be adapted as well if you apply my suggestion above
```suggestion
    for i in eachdim(mesh) 
      flux_values .= flux.(view(u_values,:,e), i, equations)
      for j in eachdim(mesh)
        StructArrays.foreachfield(mul_by_accum!(invMQrstTrW[j], rstxyzJ[i,j][1,e]), 
                                  view(du,:,e), flux_values)
      end
    end
```

ranocha commented on examples/simplicial_mesh/elixir_euler_tet_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+# !!! warning "Experimental features"
```
I think it would be nice if you could rename the new examples folder using the same template we use for the other folders. Thus, you could use two folders such as
- `examples/simplicial_2d_dg`
- `examples/simplicial_3d_dg`

to make it clear that there are both 2D and 3D examples (for DG-type methods).

sloede commented:
I started reviewing this. Not sure when it will be done, but it would be great if only necessary bug fixes could be pushed forth etime being, to reduce the rsik of one of my comments being gobbled by the changes.

jlchan commented:
@sloede Sounds good. Should leave @ranocha's comments until after you've had a chance to look at, or would you be OK if I addressed those first?

sloede reviewed (commented):
Here's a first batch of comments (I haven't seen all files yet, though)

sloede commented:
I've started to review this. Since changes pushed after I add a comment sometimes gobbles these comments, I'd prefer if only 

> @sloede Sounds good. Should leave @ranocha's comments until after you've had a chance to look at, or would you be OK if I addressed those first?

I just submitted a first round of comments, so you can go ahead and change away :-)

jlchan commented on test/test_examples_simplicial_mesh.jl:
```diff
@@ -0,0 +1,45 @@
+
+@testset "Simplicial mesh tests" begin    
+
+  @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_triangular_mesh.jl"),
+    l2 = [7.687514677136661e-5, 8.844681835345058e-5, 8.844681835367038e-5, 0.0002667915678724591], 
+    linf = [0.00023033713406972467, 0.0001967281286732181, 0.00019672812868742895, 0.0004570579510136952]
+  )
```
There's a separate elixir for the EOC convergence test. Are those usually tested too? Should there not need to be a separate elixir for EOC?

ranocha replied:
We have some convergence tests in https://github.com/trixi-framework/Trixi.jl/blob/1525846f5124e4fbd50343137e2d2157f325c78e/test/test_special_elixirs.jl#L20

jlchan commented on examples/simplicial_mesh/elixir_euler_Triangulate_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+# !!! warning "Experimental features"
```
Hm...the `Triangulate_mesh` is an unstructured mesh created by the Triangulate.jl package, which is also why it was capitalized previously. 

Maybe `eilxir_euler_triangulate_pkg_mesh.jl`?

jlchan commented on .github/workflows/ci.yml:
```diff
@@ -63,6 +63,7 @@ jobs:
           - 2d_part1
           - 2d_part2
           - 2d_part3
+          - simplices
```
Sounds good - I'll split it up into `2d_simplices` and `3d_simplices` following the setup style. 

One note - there's a single quadrilateral example within the `2d_simplices` folder. However, it uses the simplicial DG framework, so I'm including it in that test folder until I can specialize more for quads.

jlchan commented on test/test_examples_simplicial_mesh.jl:
```diff
@@ -0,0 +1,45 @@
+
+@testset "Simplicial mesh tests" begin    
+
+  @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_triangular_mesh.jl"),
+    l2 = [7.687514677136661e-5, 8.844681835345058e-5, 8.844681835367038e-5, 0.0002667915678724591], 
+    linf = [0.00023033713406972467, 0.0001967281286732181, 0.00019672812868742895, 0.0004570579510136952]
+  )
```
Does this mean the entire testset should run in under 10 seconds?

ranocha replied:
Each test should run roughly in under 10 seconds

jlchan commented on examples/simplicial_mesh/elixir_euler_Triangulate_mesh.jl:
```diff
@@ -0,0 +1,47 @@
+# !!! warning "Experimental features"
```
Tentatively renaming to `eilxir_euler_triangulate_pkg_mesh.jl`.

jlchan commented:
@ranocha I think this should be ready to merge into `dev` whenever you've had a chance to look at it.
